# Table of contents

## 01 Importing libraries

## 02 Importing data

## 03 Data wrangling

## 04 Data consistency checks

## 05 Combining data on customers with data on departments and data on orders and products

## 06 Exporting merged dataframe

# 01 Importing libraries

In [1]:
# Importing pandas, NumPy, and os
import pandas as pd
import numpy as np
import os

# 02 Importing data

In [2]:
# Defining path variable
path = r'/Users/DanielaDietmayr/Library/CloudStorage/OneDrive-Personal/2023-01-18 Instacart basket analysis'

In [3]:
# Importing data on departments
deps = pd.read_csv(os.path.join(path, '02 Data', '02 Prepared data', 'departments_wrangled.csv'), index_col = False)

In [4]:
# Printing deps
deps

Unnamed: 0       department
0            1           frozen
1            2            other
2            3           bakery
3            4          produce
4            5          alcohol
5            6    international
6            7        beverages
7            8             pets
8            9  dry goods pasta
9           10             bulk
10          11    personal care
11          12     meat seafood
12          13           pantry
13          14        breakfast
14          15     canned goods
15          16       dairy eggs
16          17        household
17          18           babies
18          19           snacks
19          20             deli
20          21          missing

In [5]:
# Renaming column 'Unnamed: 0' to department_id
deps.rename(columns = {'Unnamed: 0' : 'department_id'}, inplace = True)

In [6]:
# Printing deps again
deps

department_id       department
0               1           frozen
1               2            other
2               3           bakery
3               4          produce
4               5          alcohol
5               6    international
6               7        beverages
7               8             pets
8               9  dry goods pasta
9              10             bulk
10             11    personal care
11             12     meat seafood
12             13           pantry
13             14        breakfast
14             15     canned goods
15             16       dairy eggs
16             17        household
17             18           babies
18             19           snacks
19             20             deli
20             21          missing

In [7]:
# Importing merged data on orders and products as exported after exercise 4.8
ords_prods = pd.read_pickle(os.path.join(path, '02 Data', '02 Prepared data', 'orders_products_analysed_V2.pkl'))

In [8]:
# Checking ords_prods
ords_prods.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 24 columns):
 #   Column                         Dtype   
---  ------                         -----   
 0   product_id                     int64   
 1   product_name                   object  
 2   aisle_id                       int64   
 3   department_id                  int64   
 4   prices                         float64 
 5   order_id                       int64   
 6   user_id                        int64   
 7   order_number                   int64   
 8   order_day_of_the_week          int64   
 9   order_hour_of_day              int64   
 10  days_since_prior_order         float64 
 11  add_to_cart_order              int64   
 12  reordered                      int64   
 13  _merge                         category
 14  price_range_loc                object  
 15  Busiest_day                    object  
 16  Busiest_days                   object  
 17  Busiest_period_of_day    

In [9]:
# Checking ords_prods top rows
ords_prods.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_the_week  order_hour_of_day  \
0   3139998      138            28                      6                 11   
1   1977647      138            30                      6                 17   
2    389851      709             2                      0                 21   
3    652770      764             1                      3                 13   
4   1813452      764             3                      4                 17   

   ...    price_range_loc     Busiest_day         Busiest_days  \
0  ...  Mid-range product  Regularly busy    Average busy days   
1  ...  Mid-range product  Regularly busy    Average busy days   
2  ...  Mid-range product     Busiest day     Busiest two days   
3  ...  Mid-range product  Regularly busy  Least busy two days   
4  ...  Mid-range product      Least busy  Least busy two days   

  Busiest_period_of_day maximum_orders      loyalty_flag  \
0        Average orders             32  Regular customer   
1        Average orders             32  Regular customer   
2        Average orders              5      New customer   
3        Average orders              3      New customer   
4        Average orders              3      New customer   

  average_price_for_user spending_flag  median_days_since_prior_order  \
0               6.935811   Low spender                            8.0   
1               6.935811   Low spender                            8.0   
2               7.930208   Low spender                            6.0   
3               4.972414   Low spender                            9.0   
4               4.972414   Low spender                            9.0   

  order_frequency_flag  
0    Frequent customer  
1    Frequent customer  
2    Frequent customer  
3    Frequent customer  
4    Frequent customer  

[5 rows x 24 columns]

In [10]:
# Importing data on customers
cust_raw = pd.read_csv(os.path.join(path, '02 Data', '01 Original data', 'customers.csv'), index_col = False)

In [11]:
# Checking cust_raw
cust_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [12]:
# Checking cust_raw top rows
cust_raw.head(10)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5   133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6   152052      Chris    Walton    Male         Montana   20    1/1/2017   
7   168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8    69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9    82820      Shawn     Chung    Male        Virginia   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374  
5             2    married   49643  
6             0     single   61746  
7             0     single   63712  
8             1    married  162432  
9             2    married   32072

In [13]:
# Checking cust_raw bottom rows
cust_raw.tail(10)

user_id First Name      Surnam  Gender           STATE  Age  \
206199   179673       Adam  Villanueva    Male         Wyoming   77   
206200   192637      Kelly    Petersen  Female         Florida   47   
206201    75529      Kathy       Avila  Female   West Virginia   25   
206202   157533        Ned       Stark    Male            Utah   53   
206203   122741       Carl       Weber    Male        Michigan   24   
206204   168073       Lisa        Case  Female  North Carolina   44   
206205    49635     Jeremy     Robbins    Male          Hawaii   62   
206206   135902      Doris    Richmond  Female        Missouri   66   
206207    81095       Rose     Rollins  Female      California   27   
206208    80148    Cynthia       Noble  Female        New York   55   

       date_joined  n_dependants        fam_status  income  
206199    4/1/2020             0  divorced/widowed  162239  
206200    4/1/2020             0            single   50638  
206201    4/1/2020             2           married   56513  
206202    4/1/2020             0            single   35973  
206203    4/1/2020             2           married   73357  
206204    4/1/2020             1           married  148828  
206205    4/1/2020             3           married  168639  
206206    4/1/2020             2           married   53374  
206207    4/1/2020             1           married   99799  
206208    4/1/2020             1           married   57095

# 03 Data wrangling
Data on departments as well as orders and products has been wrangled in previous scripts. Therefore data wrangling in this scripts focuses on data on customers.

To provide the insights that part 2 of exercise 4.9 asks for the following columns are not necessary: First Name, Surnam, Gender, STATE, date_joined. Therefore, they could be dropped from the dataframe. It is however likely, that names, gender and geographical data is needed at a later stage. Therefore, only date_joined will be dropped from the dataframe by creating a new dataframe cust_raw_rel_col with only the columns relevant for the ensuing analysis. 
The remaining columns are: user_id, First Name, Surnam, Gender, STATE, Age, n_dependants, fam_status, and income. To ensure consistent naming of columns, some columns will be renamed.

Data types of the remaining columns seem accurate and do not require wrangling.

In [14]:
# Dropping unnecessary columns from cust_raw
cust_raw_rel_col = cust_raw.drop(columns = ['date_joined'])

In [15]:
# Renaming columns
cust_raw_rel_col.rename(columns = {'First Name':'first_name', 'Surnam': 'surname', 'Gender':'gender', 'STATE':'state', 'Age' : 'age'}, inplace = True)

In [16]:
# Checking cust_raw_rel_col
cust_raw_rel_col.head()

user_id first_name   surname  gender       state  age  n_dependants  \
0    26711    Deborah  Esquivel  Female    Missouri   48             3   
1    33890   Patricia      Hart  Female  New Mexico   36             0   
2    65803    Kenneth    Farley    Male       Idaho   35             2   
3   125935   Michelle     Hicks  Female        Iowa   40             0   
4   130797        Ann   Gilmore  Female    Maryland   26             1   

  fam_status  income  
0    married  165665  
1     single   59285  
2    married   99568  
3     single   42049  
4    married   40374

In [18]:
# Checking dimensions of cust_raw_rel_col
cust_raw_rel_col.shape

(206209, 9)

# 04 Data consistency checks
Consistency of data on departments as well as orders and products has been checked in previous scripts. Therefore data consistency checks in this scripts focuses on data on customers.

From the brief checks of top and bottom rows of the customer data, there seem to be no missing values. This will be double-checked by investigation of the remaining columns. 

Descriptive statistics will be run on all columns to spot unexpected values.

From the brief checks above, there seem to be no mixed-type columns in the customer data. This will be double-checked by in-depth investigation. 

Last, but not least, customer data will be checked for duplicates. 

In [19]:
# Checking for missing values with isna
cust_raw_rel_col.isna().sum()

user_id             0
first_name      11259
surname             0
gender              0
state               0
age                 0
n_dependants        0
fam_status          0
income              0
dtype: int64

Column 'first name' contains missing values. However, as it won't be needed for the analysis, it remains unadressed.

In [20]:
# Checking for unexpected values
cust_raw_rel_col.describe()

user_id            age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [21]:
# Getting full row for observation with maximum income = 593901
cust_raw_rel_col.loc[cust_raw_rel_col['income']==593901]

user_id first_name surname gender       state  age  n_dependants  \
21726    58596     Thomas   Cohen   Male  Washington   75             2   

      fam_status  income  
21726    married  593901

The only descriptive statistic that seems suspect is the maximum income of 593.901$. Checking the full observation shows that the user with this income is 75 years old, married and living in a household with two further persons. 
While most people can only dream of such a high income, it is not impossible and is therefore kept in the dataset. 

In [22]:
# Checking for mixed-type columns in cust_raw_rel_col with for loop
for col in cust_raw_rel_col.columns.tolist():
    weird = (cust_raw_rel_col[[col]].applymap(type) != cust_raw_rel_col[[col]].iloc[0].apply(type)).any(axis=1)
    if len (cust_raw_rel_col[weird]) > 0:
        print(col)

first_name


In [25]:
# Addressing mixed-type column
cust_raw_rel_col['first_name'] = cust_raw_rel_col['first_name'].astype('str')

In [26]:
# Checking for mixed-type columns in cust_raw_rel_col again
for col in cust_raw_rel_col.columns.tolist():
    weird = (cust_raw_rel_col[[col]].applymap(type) != cust_raw_rel_col[[col]].iloc[0].apply(type)).any(axis=1)
    if len (cust_raw_rel_col[weird]) > 0:
        print(col)

No mixed-type columns in cust_raw_rel_col remaining.

In [27]:
# Checking for duplicates in cust_raw_rel_col
cust_raw_rel_col_dups = cust_raw_rel_col[cust_raw_rel_col.duplicated()]

In [28]:
# Printing top row of cust_raw_rel_col_dups
cust_raw_rel_col_dups.head()

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, n_dependants, fam_status, income]
Index: []

There are no full duplicates in the customer dataset. 

# 05 Combining data on customers with data on departments and data on orders and products
Before merging the three dataframes, the key columns required for the operations are checked. Key columns are: department_id in deps, department_id and user_id in ords_prods, and user_id in cust_raw_rel_col.

Additionally, the _merge flag from a previous merging operation will be removed from the data on orders and products.

Furthermore, the dimensions of all dataframes are checked to provide for accuracy checks after merging.

As for the purpose of the analysis in part 2 of this exercise a fully matching dataset is needed, an inner join will be used to merge the dataset. To check whether the datasets actually are fully matching, a test-version of the merged dataframe will be created with an outer join. This test-version will be compared to the merged dataframe created with an inner join.

Finally, the following checks will be run before exporting the merged dataframe:
retrieving basic info about the merged dataframe;
displaying the top and bottom rows;
checking for missing values.

In [29]:
# Checking datatype of department_id in deps
deps['department_id'].dtype

dtype('int64')

In [30]:
# Checking datatypes of department_id and user_id in ords_prods
ords_prods[['department_id', 'user_id']].dtypes

department_id    int64
user_id          int64
dtype: object

In [31]:
# Checking datatype of user_id in cust_raw_rel_col
cust_raw_rel_col['user_id'].dtype

dtype('int64')

In [32]:
# Dropping _merge column from ords_prods
ords_prods_dropped = ords_prods.drop(columns = ['_merge'])

In [33]:
# Checking dimensions of deps
deps.shape

(21, 2)

In [34]:
# Checking dimensions of ords_prods
ords_prods_dropped.shape

(32404859, 23)

In [35]:
# Checking dimensions of cust_raw_rel_col
cust_raw_rel_col.shape

(206209, 9)

If the datasets match fully, the merged dataframe should have dimensions of 32404859, 32.

In [36]:
# Merging deps, ords_prods_dropped and cust_raw_rel_col
df_merged = deps.merge(ords_prods_dropped, on = 'department_id').merge(cust_raw_rel_col, on = 'user_id')

In [37]:
# Checking dimensions of df_merged
df_merged.shape

(32404859, 32)

In [38]:
# Merging deps, ords_prods_dropped and cust_raw_rel_col with outer join
df_merged_test = deps.merge(ords_prods_dropped, on = 'department_id', how = 'outer').merge(cust_raw_rel_col, on = 'user_id', how = 'outer')

In [39]:
# Checking dimensions of df_merged_test
df_merged_test.shape

(32404859, 32)

The checks confirm that the initial three dataframes are fully matching. 

In [40]:
# Checking basic info on df_merged
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 32 columns):
 #   Column                         Dtype  
---  ------                         -----  
 0   department_id                  int64  
 1   department                     object 
 2   product_id                     int64  
 3   product_name                   object 
 4   aisle_id                       int64  
 5   prices                         float64
 6   order_id                       int64  
 7   user_id                        int64  
 8   order_number                   int64  
 9   order_day_of_the_week          int64  
 10  order_hour_of_day              int64  
 11  days_since_prior_order         float64
 12  add_to_cart_order              int64  
 13  reordered                      int64  
 14  price_range_loc                object 
 15  Busiest_day                    object 
 16  Busiest_days                   object 
 17  Busiest_period_of_day          object 
 18  

In [41]:
# Changing pandas options to display all columns
pd.set_option('display.max_columns', None)

In [42]:
# Checking top rows of df_merged
df_merged.head()

department_id department  product_id  \
0              1     frozen           4   
1              1     frozen         311   
2              1     frozen         311   
3              1     frozen        2452   
4              1     frozen        2452   

                                        product_name  aisle_id  prices  \
0  Smart Ones Classic Favorites Mini Rigatoni Wit...        38    10.5   
1           Naturals Savory Turkey Breakfast Sausage        52     3.5   
2           Naturals Savory Turkey Breakfast Sausage        52     3.5   
3                           Naturals Chicken Nuggets       129    11.0   
4                           Naturals Chicken Nuggets       129    11.0   

   order_id  user_id  order_number  order_day_of_the_week  order_hour_of_day  \
0   2563211     1677             1                      5                 13   
1   2135223     1677            18                      4                 22   
2   2660694     1677            19                      6                 13   
3   1050238     1677            14                      6                  0   
4   2149917     1677            16                      5                 18   

   days_since_prior_order  add_to_cart_order  reordered    price_range_loc  \
0                     0.0                  9          0  Mid-range product   
1                     4.0                 10          0  Low-range product   
2                    16.0                 15          1  Low-range product   
3                    10.0                 16          0  Mid-range product   
4                    11.0                  9          1  Mid-range product   

      Busiest_day         Busiest_days Busiest_period_of_day  maximum_orders  \
0  Regularly busy    Average busy days        Average orders              21   
1      Least busy  Least busy two days        Average orders              21   
2  Regularly busy    Average busy days        Average orders              21   
3  Regularly busy    Average busy days        Average orders              21   
4  Regularly busy    Average busy days        Average orders              21   

       loyalty_flag  average_price_for_user spending_flag  \
0  Regular customer                 7.47101   Low spender   
1  Regular customer                 7.47101   Low spender   
2  Regular customer                 7.47101   Low spender   
3  Regular customer                 7.47101   Low spender   
4  Regular customer                 7.47101   Low spender   

   median_days_since_prior_order order_frequency_flag first_name surname  \
0                           13.0     Regular customer        Tom   White   
1                           13.0     Regular customer        Tom   White   
2                           13.0     Regular customer        Tom   White   
3                           13.0     Regular customer        Tom   White   
4                           13.0     Regular customer        Tom   White   

  gender state  age  n_dependants fam_status  income  
0   Male  Utah   57             2    married  101071  
1   Male  Utah   57             2    married  101071  
2   Male  Utah   57             2    married  101071  
3   Male  Utah   57             2    married  101071  
4   Male  Utah   57             2    married  101071

In [43]:
# Checking bottom rows of df_merged
df_merged.tail()

department_id department  product_id         product_name  aisle_id  \
32404854             20       deli       35652  Prosciutto Di Parma        96   
32404855             20       deli       35652  Prosciutto Di Parma        96   
32404856             20       deli       41588            Firm Tofu        14   
32404857             20       deli       41588            Firm Tofu        14   
32404858             20       deli       41588            Firm Tofu        14   

          prices  order_id  user_id  order_number  order_day_of_the_week  \
32404854     5.4   1622762    40200             2                      4   
32404855     5.4    258797    40200             3                      3   
32404856    11.1    812660    94318             1                      5   
32404857    11.1   1000709    94318             2                      2   
32404858    11.1     51828    94318             3                      1   

          order_hour_of_day  days_since_prior_order  add_to_cart_order  \
32404854                 19                    30.0                  1   
32404855                  8                    30.0                  1   
32404856                 15                     0.0                  1   
32404857                 12                    11.0                  1   
32404858                 12                     6.0                  1   

          reordered    price_range_loc     Busiest_day         Busiest_days  \
32404854          1  Mid-range product      Least busy  Least busy two days   
32404855          1  Mid-range product  Regularly busy  Least busy two days   
32404856          0  Mid-range product  Regularly busy    Average busy days   
32404857          1  Mid-range product  Regularly busy    Average busy days   
32404858          1  Mid-range product  Regularly busy     Busiest two days   

         Busiest_period_of_day  maximum_orders  loyalty_flag  \
32404854        Average orders               3  New customer   
32404855        Average orders               3  New customer   
32404856        Average orders               3  New customer   
32404857        Average orders               3  New customer   
32404858        Average orders               3  New customer   

          average_price_for_user spending_flag  median_days_since_prior_order  \
32404854                     5.4   Low spender                           30.0   
32404855                     5.4   Low spender                           30.0   
32404856                    11.1  High spender                            6.0   
32404857                    11.1  High spender                            6.0   
32404858                    11.1  High spender                            6.0   

           order_frequency_flag first_name     surname  gender      state  \
32404854  Non-frequent customer      Frank     Herring    Male     Hawaii   
32404855  Non-frequent customer      Frank     Herring    Male     Hawaii   
32404856      Frequent customer    Barbara  Harrington  Female  Louisiana   
32404857      Frequent customer    Barbara  Harrington  Female  Louisiana   
32404858      Frequent customer    Barbara  Harrington  Female  Louisiana   

          age  n_dependants fam_status  income  
32404854   45             1    married   55272  
32404855   45             1    married   55272  
32404856   48             1    married  115470  
32404857   48             1    married  115470  
32404858   48             1    married  115470

In [44]:
# Checking for missing values
df_merged.isna().sum()

department_id                    0
department                       0
product_id                       0
product_name                     0
aisle_id                         0
prices                           0
order_id                         0
user_id                          0
order_number                     0
order_day_of_the_week            0
order_hour_of_day                0
days_since_prior_order           0
add_to_cart_order                0
reordered                        0
price_range_loc                  0
Busiest_day                      0
Busiest_days                     0
Busiest_period_of_day            0
maximum_orders                   0
loyalty_flag                     0
average_price_for_user           0
spending_flag                    0
median_days_since_prior_order    0
order_frequency_flag             0
first_name                       0
surname                          0
gender                           0
state                            0
age                 

# 06 Exporting merged dataframe

In [45]:
# Exporting df_merged
df_merged.to_pickle(os.path.join(path, '02 Data', '02 Prepared data', 'ords_prods_cust_deps.pkl'))